atc/exec/set_pipeline_step.go

```diff
@@ -0,0 +1,243 @@
+		}
+
+		step.delegate.Finished(logger, false)
+		return nil
+	}
+
+	team := step.teamFactory.GetByID(step.metadata.TeamID)
+
+	fromVersion := db.ConfigVersion(0)
+	pipeline, found, err := team.Pipeline(step.plan.Name)
+	if err != nil {
+		return nil
+	}
+
+	var existingConfig atc.Config
+	if !found {
+		existingConfig = atc.Config{}
+	} else {
+		fromVersion = pipeline.ConfigVersion()
+		existingConfig, err = pipeline.Config()
+		if err != nil {
+			return err
+		}
+	}
+
+	diffExists := existingConfig.Diff(stdout, *atcConfig)
+	if !diffExists {
+		logger.Debug("no-diff")
+
+		fmt.Fprintf(stdout, "No diff found.\n")
+		step.succeeded = true
+		step.delegate.Finished(logger, true)
+		return nil
+	}
+
+	fmt.Fprintf(stdout, "Updating the pipeline.\n")
+	pipeline, _, err = team.SavePipeline(step.plan.Name, *atcConfig, fromVersion, false)
+	if err != nil {
+		return err
+	}
+
+	fmt.Fprintf(stdout, "Done successfully.\n")
+	logger.Info("saved-pipeline", lager.Data{"team": team.Name(), "pipeline": pipeline.Name()})
+	step.succeeded = true
+	step.delegate.Finished(logger, true)
+
+	return nil
+}
+
+func (step *SetPipelineStep) Succeeded() bool {
+	return step.succeeded
+}
+
+type setPipelineSource struct {
+	ctx    context.Context
+	logger lager.Logger
+	repo   *artifact.Repository
+	step   *SetPipelineStep
+}
+
+// streamInBytes streams a file from other resource and returns a byte array.
+func (s setPipelineSource) streamInBytes(path string) ([]byte, error) {
+	stream, err := s.repo.StreamFile(s.ctx, s.logger, path)
+	if err != nil {
+		return nil, err
+	}
+	defer stream.Close()
+
+	byteConfig, err := ioutil.ReadAll(stream)
+	if err != nil {
+		return nil, err
+	}
+
+	return byteConfig, nil
+}
+
+// FetchConfig streams pipeline configure file and var files from other resources
+// and construct an atc.Config object.
+func (s setPipelineSource) FetchConfig() (*atc.Config, error) {
+	config, err := s.streamInBytes(s.step.plan.File)
+	if err != nil {
+		return nil, err
+	}
+
+	staticVarss := []vars.Variables{}
+	if len(s.step.plan.Vars) > 0 {
+		staticVarss = append(staticVarss, vars.StaticVariables(s.step.plan.Vars))
+	}
+	for _, lvf := range s.step.plan.VarFiles {
+		bytes, err := s.streamInBytes(lvf)
+		if err != nil {
+			return nil, err
+		}
+
+		sv := vars.StaticVariables{}
+		err = yaml.Unmarshal(bytes, &sv)
+		if err != nil {
+			return nil, err
+		}
+
+		staticVarss = append(staticVarss, sv)
+	}
+
+	if len(staticVarss) > 0 {
+		config, err = vars.NewTemplateResolver(config, staticVarss).Resolve(false, false)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	atcConfig := atc.Config{}
+	err = yaml.Unmarshal(config, &atcConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &atcConfig, nil
+}
+
+func (s setPipelineSource) Validate() error {
+	if s.step.plan.File == "" {
+		return errors.New("file is not specified")
+	}
+
+	return nil
+}
+
+func (s setPipelineSource) EvaluatePlan(variables vars.Variables) error {
```

Comment:
This is interesting; what's the use case for evaluating all of these? I only ask because we don't do it for any of the other steps.

I don't think you need to do it for the `vars:` since any vars contained within will be recursively evaluated when resolving the pipeline config.

e.g.:

```yaml
set_pipeline: my-pipeline
vars:
  foo: ((bar))
```

In the pipeline, `((foo))` will resolve to `((bar))` which will then be resolved again to its value. Whereas here it's evaluated ahead-of-time so `((foo))` will resolve directly to the value.

So at least for `vars:` I don't think this is necessary as things should resolve the same. And aside from `vars:` I don't see a whole lot of value in evaluating the vars in `file:` or `var_files:`. :thinking:

Reply:
> This is interesting; what's the use case for evaluating all of these? I only ask because we don't do it for any of the other steps.
> 

I'm a little confused. I think all other steps evaluate `source` and `params`, for example:

```yml
resource:
- name: mygit
   type: git
   source:
       ...
       private_key: ((chaol-gitlab-private-key))

jobs:
- name: myjob
   plan:
   - task: mytask
      ...
      run:
          path: echo
          args:
          - Hello ((name))

    - task: othertask
       file: mygit/((path)/file.yml
```

Maybe you meant we don't evaluate fields other than `source` and `params`. But here I consider `file`, `var_files` etc. as params of `set_pipeline` step though I put it at top level per your suggestion.

> I don't think you need to do it for the `vars:` since any vars contained within will be recursively evaluated when resolving the pipeline config.
> 

For evaluating `vars` is, I initially thought for the self reset use case. Say, when I initially set the pipeline by `fly`, I run the command:

```bash
$ fly -t some sp -c pipeline.yml -p name -v foo=bar
```

Then self reset would still need to pass `foo`, like:

```yml

jobs:
- name: self-reset
  plan:
  - get: mygit
  - set-pipeline: self
     file: mygit/pipeline.yml
     vars:
        foo: ((foo))

- name: normal-job
   plan:
   - task:
      ...
       run: echo ((foo))
```

But, yes, you are right, in `noral-job`, `((foo))` has been interpolated during `fly sp`, the `set-pipeline` step even no need to have `vars` in this case. So I'll remove `vars` from evaluation.

I'll keep evaluation for `file` and `var_files`, which is required for pipeline template.

Reply:
There are two ways for vars to be expanded: static expansion done at `set-pipeline` time, which is non-semantic and happens client-side across the template, and dynamic expansion done at runtime, which is only done for specific attributes which support it, i.e. a resource's `source` and `params` (not general `params` - specifically `params` which deal with resources and `params` which are given to tasks as env vars).

The `vars` field doesn't need to be evaluated because any vars given at `set-pipeline` time will take precedence, and any vars being provided at runtime will just be recursively evaluated once they're pulled into the pipeline template (from `vars:`).

My only concern is that by evaluating all of the other fields this step is inconsistent with the other steps which don't. You don't need to explicitly implement evaluation if you're providing the vars statically at `set-pipeline` time, which it seems like would be the only use case for `file` and `var_files`.

For example, a task step won't evaluate any vars in its `file:` field. This example would only work if `path` is provided at `set-pipeline` time:

```yaml
plan:
- task: othertask
  file: mygit/((path))/file.yml
```

...but with this change, `file:` would be able to expand at runtime on the `set-pipeline` step, but there doesn't seem to be a strong reason for this behavior.

Reply:
@vito Maybe I over considered potential use case here. I agree to keep consistent behavior of `file`. Let me remove evaluation of `file` and `var_files` for now.